File: backend/app/schemas/company.py
import uuid
from datetime import datetime
from pydantic import BaseModel, field_validator

class CreateCompanyRequest(BaseModel):
    """Ce trimite SUPER_ADMIN când creează o companie nouă."""
    name: str
    slug: str
    plan: str = "FREE"
    max_vehicles: int = 10
    max_users: int = 20

    @field_validator("slug")
    @classmethod
    def slug_alphanumeric(cls, value: str) -> str:
        if not value.replace("-", "").replace("_", "").isalnum():
            raise ValueError("Slug-ul poate conține doar litere, cifre, - și _")
        return value.lower()

class UpdateCompanyRequest(BaseModel):
    """Ce trimite SUPER_ADMIN când actualizează o companie."""
    name: str | None = None
    is_active: bool | None = None
    plan: str | None = None
    max_vehicles: int | None = None
    max_users: int | None = None


class CompanyResponse(BaseModel):
    id: uuid.UUID
    name: str
    slug: str
    is_active: bool
    plan: str
    max_vehicles: int | None
    max_users: int | None
    managers_count: int = 0
    users_count: int = 0
    dispatchers_count: int = 0
    drivers_count: int = 0
    clients_count: int = 0
    vehicles_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PublicCompanyResponse(BaseModel):
    name: str
    slug: str

    model_config = {"from_attributes": True}


class CompanyStatsResponse(BaseModel):
    """Statistici pentru o companie specifică."""
    company_id: uuid.UUID
    company_name: str
    is_active: bool
    total_users: int
    total_managers: int
    total_dispatchers: int
    total_drivers: int
    total_clients: int
    total_vehicles: int
    total_orders: int
    total_trips: int
    plan: str
    max_vehicles: int | None
    max_users: int | None

File: frontend/src/app/core/services/company.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { AuthService } from './auth';

export interface CreateCompanyRequest {
  name: string;
  slug: string;
  plan: string;
  max_vehicles: number;
  max_users: number;
  depot_county?: string | null;
  depot_city?: string | null;
  depot_street?: string | null;
  depot_number?: string | null;
}

export interface UpdateCompanyRequest {
  name?: string;
  is_active?: boolean;
  plan?: string;
  max_vehicles?: number;
  max_users?: number;
  depot_county?: string | null;
  depot_city?: string | null;
  depot_street?: string | null;
  depot_number?: string | null;
}

export interface CompanyResponse {
  id: string;
  name: string;
  slug: string;
  is_active: boolean;
  plan: string;
  max_vehicles: number | null;
  max_users: number | null;
  depot_county: string | null;
  depot_city: string | null;
  depot_street: string | null;
  depot_number: string | null;
  depot_lat: number | null;
  depot_lon: number | null;
  managers_count: number;
  users_count: number;
  dispatchers_count: number;
  drivers_count: number;
  clients_count: number;
  vehicles_count: number;
  created_at: string;
  updated_at: string;
}

export interface CompanyStatsResponse {
  company_id: string;
  company_name: string;
  is_active: boolean;
  total_users: number;
  total_managers: number;
  total_dispatchers: number;
  total_drivers: number;
  total_clients: number;
  total_vehicles: number;
  total_orders: number;
  total_trips: number;
  plan: string;
  max_vehicles: number | null;
  max_users: number | null;
}

export interface PublicCompanyResponse {
  name: string;
  slug: string;
}

@Injectable({
  providedIn: 'root',
})
export class CompanyService {
  private readonly apiUrl = 'http://127.0.0.1:8000/companies';

  constructor(
    private http: HttpClient,
    private authService: AuthService,
  ) {}

  listCompanies(): Observable<CompanyResponse[]> {
    return this.http.get<CompanyResponse[]>(`${this.apiUrl}/`, {
      headers: this.authService.getAuthHeaders(),
    });
  }

  listSignupCompanies(): Observable<PublicCompanyResponse[]> {
    return this.http.get<PublicCompanyResponse[]>(`${this.apiUrl}/public/signup-options`);
  }

  createCompany(data: CreateCompanyRequest): Observable<CompanyResponse> {
    return this.http.post<CompanyResponse>(`${this.apiUrl}/`, data, {
      headers: this.authService.getAuthHeaders(),
    });
  }

  updateCompany(companyId: string, data: UpdateCompanyRequest): Observable<CompanyResponse> {
    return this.http.patch<CompanyResponse>(`${this.apiUrl}/${companyId}`, data, {
      headers: this.authService.getAuthHeaders(),
    });
  }

  getCompanyStats(companyId: string): Observable<CompanyStatsResponse> {
    return this.http.get<CompanyStatsResponse>(`${this.apiUrl}/${companyId}/stats`, {
      headers: this.authService.getAuthHeaders(),
    });
  }

  activateCompany(companyId: string): Observable<CompanyResponse> {
    return this.http.patch<CompanyResponse>(
      `${this.apiUrl}/${companyId}/activate`,
      {},
      {
        headers: this.authService.getAuthHeaders(),
      },
    );
  }

  deactivateCompany(companyId: string): Observable<CompanyResponse> {
    return this.http.patch<CompanyResponse>(
      `${this.apiUrl}/${companyId}/deactivate`,
      {},
      {
        headers: this.authService.getAuthHeaders(),
      },
    );
  }
}
